File: alchem-link/src/alchem_link/rpc.py
"""A small, honest JSON-RPC client built on the standard library.

No ``requests``, no ``web3`` — one stdlib HTTP call plus JSON. Two details are not
obvious and both cost real debugging time if you meet them cold:

* **User-Agent.** Several public RPC providers reject Python's default
  ``Python-urllib/3.x`` with a 403, which reads as "the chain is down". This client
  always sends a real one.
* **Batching.** JSON-RPC 2.0 lets you POST an *array* of requests and get an array back,
  which turns twenty feed reads into one round trip. Providers may return the responses
  in any order, so results are re-sorted by request id rather than by position — a
  mistake that silently pairs each answer with the wrong question.
"""
from __future__ import annotations

import json
import time
import urllib.error
import urllib.request
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional, Sequence, Tuple

from .abi import (
    SELECTOR_DECIMALS,
    SELECTOR_DESCRIPTION,
    SELECTOR_LATEST_ROUND_DATA,
)
from .networks import DEFAULT_NETWORK, Endpoint, resolve_endpoint

DEFAULT_TIMEOUT = 15.0
DEFAULT_RETRIES = 2

#: Providers vary, but a few hundred sub-requests is where batch payloads start getting
#: rejected outright. Chunking below that keeps a large read from failing as one unit.
MAX_BATCH_SIZE = 100


class RpcError(RuntimeError):
    """A JSON-RPC level error — the node answered, and the answer was an error."""


class RpcTransportError(RuntimeError):
    """The node could not be reached, or did not answer in time.

    ``retryable`` is False for failures that cannot improve on a second attempt —
    a 4xx other than 429, for instance. Retrying those just burns rate limit.
    """

    def __init__(self, message: str, retryable: bool = True) -> None:
        super().__init__(message)
        self.retryable = retryable


@dataclass
class RpcCallResult:
    result: Any
    elapsed_ms: float


@dataclass
class BatchOutcome:
    """One entry of a batch response. Either ``result`` is set, or ``error`` is."""
    result: Any = None
    error: Optional[str] = None

    @property
    def ok(self) -> bool:
        return self.error is None


@dataclass
class RpcStats:
    """Cheap counters, so `doctor` and the TUI can show what a run actually cost."""
    requests: int = 0
    http_posts: int = 0
    retries: int = 0
    total_ms: float = 0.0
    #: Contract reads executed *inside* a Multicall3 aggregate. These never become
    #: JSON-RPC requests at all, so counting only `requests` understates the saving.
    batched_reads: int = 0

    @property
    def logical_reads(self) -> int:
        """What the same work would have cost as one request each."""
        return self.requests + self.batched_reads

    def as_dict(self) -> Dict[str, Any]:
        return {
            "requests": self.requests,
            "batched_reads": self.batched_reads,
            "logical_reads": self.logical_reads,
            "http_posts": self.http_posts,
            "retries": self.retries,
            "total_ms": round(self.total_ms, 1),
            "round_trips_saved": max(0, self.logical_reads - self.http_posts),
        }


class RpcClient:
    """Minimal JSON-RPC 2.0 client.

    Retries only transport failures. A JSON-RPC error is a real answer from the node —
    retrying a bad request just wastes the caller's rate limit.
    """

    def __init__(
        self,
        endpoint: Endpoint,
        timeout: float = DEFAULT_TIMEOUT,
        retries: int = DEFAULT_RETRIES,
        user_agent: str = "alchem-link",
    ) -> None:
        self.endpoint = endpoint
        self.timeout = timeout
        self.retries = max(0, retries)
        self.user_agent = user_agent
        self.stats = RpcStats()
        self._request_id = 0
        #: Cached Multicall3 probe. One round trip per client, not per batch.
        self.multicall3_supported: Optional[bool] = None

    # ── transport ────────────────────────────────────────────────────────────

    def _attempt(self, payload: bytes) -> Any:
        """Exactly one HTTP round trip. Retry policy lives in `_post`, not here."""
        request = urllib.request.Request(
            self.endpoint.url,
            data=payload,
            headers={
                "Content-Type": "application/json",
                # Required: some providers 403 the default urllib agent outright.
                "User-Agent": self.user_agent,
                "Accept": "application/json",
            },
        )
        try:
            with urllib.request.urlopen(request, timeout=self.timeout) as response:
                return json.loads(response.read().decode("utf-8"))
        except urllib.error.HTTPError as exc:
            detail = ""
            try:
                detail = exc.read().decode("utf-8", "replace")[:200]
            except Exception:  # pragma: no cover - diagnostic best effort
                pass
            # 4xx will not become 2xx on a retry; fail fast and say why. 429 is the
            # exception — backing off is exactly the right response there.
            fatal = 400 <= exc.code < 500 and exc.code != 429
            raise RpcTransportError(
                f"HTTP {exc.code} from {self.endpoint.redacted()}"
                f"{': ' + detail if detail else ''}",
                retryable=not fatal,
            ) from exc
        except json.JSONDecodeError as exc:
            raise RpcTransportError(
                f"{self.endpoint.redacted()} returned a non-JSON body", retryable=False
            ) from exc
        except (urllib.error.URLError, TimeoutError, OSError) as exc:
            raise RpcTransportError(f"{self.endpoint.redacted()}: {exc}") from exc

    def _post(self, payload: bytes) -> Any:
        last_error: Optional[Exception] = None
        started = time.perf_counter()
        for attempt in range(self.retries + 1):
            try:
                body = self._attempt(payload)
                self.stats.http_posts += 1
                self.stats.total_ms += (time.perf_counter() - started) * 1000
                return body
            except RpcTransportError as exc:
                if not exc.retryable:
                    self.stats.http_posts += 1
                    raise
                last_error = exc
                if attempt < self.retries:
                    self.stats.retries += 1
                    time.sleep(0.25 * (attempt + 1))
        self.stats.http_posts += 1
        raise RpcTransportError(
            f"could not reach {self.endpoint.redacted()} after "
            f"{self.retries + 1} attempt(s): {last_error}"
        )

    def _next_id(self) -> int:
        self._request_id += 1
        return self._request_id

    def call(self, method: str, params: Optional[List[Any]] = None) -> RpcCallResult:
        request_id = self._next_id()
        self.stats.requests += 1
        payload = json.dumps(
            {"jsonrpc": "2.0", "id": request_id, "method": method, "params": params or []}
        ).encode("utf-8")

        started = time.perf_counter()
        body = self._post(payload)
        elapsed_ms = (time.perf_counter() - started) * 1000

        if isinstance(body, dict) and body.get("error"):
            raise RpcError(f"{method} failed: {_error_text(body['error'])}")
        if not isinstance(body, dict) or "result" not in body:
            raise RpcError(f"{method} returned no result field")
        return RpcCallResult(result=body["result"], elapsed_ms=elapsed_ms)

    def batch(self, requests: Sequence[Tuple[str, List[Any]]]) -> List[BatchOutcome]:
        """Send many JSON-RPC requests in as few HTTP round trips as possible.

        Per-request failures come back as :class:`BatchOutcome` with ``error`` set rather
        than raising, because the whole point is that one bad call should not discard the
        other ninety-nine. Transport failures still raise — nothing came back at all.
        """
        if not requests:
            return []

        outcomes: List[BatchOutcome] = []
        for start in range(0, len(requests), MAX_BATCH_SIZE):
            chunk = list(requests[start:start + MAX_BATCH_SIZE])
            ids = [self._next_id() for _ in chunk]
            self.stats.requests += len(chunk)
            payload = json.dumps(
                [
                    {"jsonrpc": "2.0", "id": rid, "method": method, "params": params or []}
                    for rid, (method, params) in zip(ids, chunk)
                ]
            ).encode("utf-8")

            body = self._post(payload)

            # A provider that does not support batching may answer a single object, or
            # an error object, instead of an array. Fall back rather than crash.
            if isinstance(body, dict):
                detail = _error_text(body.get("error")) if body.get("error") else "not an array"
                outcomes.extend(self._sequential(chunk, reason=detail))
                continue
            if not isinstance(body, list):
                outcomes.extend(self._sequential(chunk, reason="unrecognised batch reply"))
                continue

            # Responses may arrive in any order; pair them by id, never by position.
            by_id = {entry.get("id"): entry for entry in body if isinstance(entry, dict)}
            for rid in ids:
                entry = by_id.get(rid)
                if entry is None:
                    outcomes.append(BatchOutcome(error="no response for this request id"))
                elif entry.get("error"):
                    outcomes.append(BatchOutcome(error=_error_text(entry["error"])))
                elif "result" not in entry:
                    outcomes.append(BatchOutcome(error="response had no result field"))
                else:
                    outcomes.append(BatchOutcome(result=entry["result"]))
        return outcomes

    def _sequential(
        self, chunk: Sequence[Tuple[str, List[Any]]], reason: str
    ) -> List[BatchOutcome]:
        """One-at-a-time fallback for endpoints that do not honour batch payloads."""
        out: List[BatchOutcome] = []
        for method, params in chunk:
            try:
                out.append(BatchOutcome(result=self.call(method, params).result))
            except RpcError as exc:
                out.append(BatchOutcome(error=str(exc)))
            except RpcTransportError as exc:
                out.append(BatchOutcome(error=f"{reason}; and then: {exc}"))
        return out

    # ── convenience wrappers ─────────────────────────────────────────────────

    def block_number(self) -> int:
        return int(self.call("eth_blockNumber").result, 16)

    def chain_id(self) -> int:
        return int(self.call("eth_chainId").result, 16)

    def gas_price_wei(self) -> int:
        return int(self.call("eth_gasPrice").result, 16)

    def eth_call(self, to: str, data: str, block: str = "latest") -> str:
        return self.call("eth_call", [{"to": to, "data": data}, block]).result

    def get_code(self, address: str, block: str = "latest") -> str:
        return self.call("eth_getCode", [address, block]).result

    def has_code(self, address: str) -> bool:
        """True when something is deployed at ``address``.

        An ``eth_call`` to an empty address returns ``0x`` rather than reverting, so a
        typo'd address looks exactly like a function that returned nothing.
        """
        try:
            return len(self.get_code(address)) > 2
        except (RpcError, RpcTransportError):
            return False

    def get_block(self, block: str = "latest", full: bool = False) -> Dict[str, Any]:
        return self.call("eth_getBlockByNumber", [block, full]).result

    def fee_history(
        self, blocks: int = 20, newest: str = "latest", percentiles: Optional[List[float]] = None
    ) -> Dict[str, Any]:
        return self.call(
            "eth_feeHistory", [hex(blocks), newest, percentiles or [10.0, 50.0, 90.0]]
        ).result

    def read_aggregator(self, address: str) -> Dict[str, str]:
        """Fetch the three aggregator fields a price read needs, as raw hex.

        Batched into one HTTP round trip. Kept as a method so callers that hold a client
        do not need to know about :mod:`alchem_link.multicall`.
        """
        selectors = [
            ("latest_round_data", SELECTOR_LATEST_ROUND_DATA),
            ("decimals", SELECTOR_DECIMALS),
            ("description", SELECTOR_DESCRIPTION),
        ]
        outcomes = self.batch(
            [("eth_call", [{"to": address, "data": data}, "latest"]) for _, data in selectors]
        )
        out: Dict[str, str] = {}
        for (name, _), outcome in zip(selectors, outcomes):
            if not outcome.ok:
                raise RpcError(f"{name}() on {address}: {outcome.error}")
            out[name] = outcome.result
        return out


def _error_text(err: Any) -> str:
    if isinstance(err, dict):
        code = err.get("code")
        message = err.get("message", "unknown error")
        return f"{message}{f' [{code}]' if code is not None else ''}"
    return str(err)


def client_for(
    network: str = DEFAULT_NETWORK,
    rpc_url: Optional[str] = None,
    timeout: float = DEFAULT_TIMEOUT,
    user_agent: str = "alchem-link",
) -> RpcClient:
    """Build a client for a network, resolving the endpoint from args and environment."""
    return RpcClient(
        resolve_endpoint(network=network, rpc_url=rpc_url),
        timeout=timeout,
        user_agent=user_agent,
    )


def gwei(wei: int) -> float:
    return wei / 1_000_000_000

File: web/lib/alchem/rpc.ts
// Server-side JSON-RPC client — a port of `alchem_link/rpc.py` on `fetch`.
//
// Differences from the Python client, both forced by the setting rather than taste:
//
//   * Calls are **batched** into one HTTP request. The CLI reads six feeds with eighteen
//     sequential round trips and nobody minds; a web panel polling on a timer would burn
//     a public endpoint's rate limit doing the same. A provider that rejects batch gets
//     detected and falls back to individual calls.
//   * `decimals()` and `description()` are cached per aggregator. They are immutable on
//     the contract, so re-reading them every poll buys nothing. Only `latestRoundData()`
//     is actually re-fetched.

import {
  SELECTOR_DECIMALS,
  SELECTOR_DESCRIPTION,
  SELECTOR_LATEST_ROUND_DATA,
} from './abi'
import { redact, type Endpoint } from './endpoint'

export const DEFAULT_TIMEOUT_MS = 12_000
export const DEFAULT_RETRIES = 2

/** The node answered, and the answer was an error. */
export class RpcError extends Error {
  constructor(message: string) {
    super(message)
    this.name = 'RpcError'
  }
}

/** The node could not be reached, or did not answer in time. */
export class RpcTransportError extends Error {
  readonly retryable: boolean
  constructor(message: string, retryable = true) {
    super(message)
    this.name = 'RpcTransportError'
    this.retryable = retryable
  }
}

interface RpcRequest {
  method: string
  params?: unknown[]
}

interface RpcEnvelope {
  id: number
  result?: unknown
  error?: { code?: number; message?: string }
}

/** One call's outcome inside a batch: a result, or the reason that single call failed. */
export type Settled =
  | { ok: true; result: unknown }
  | { ok: false; error: string }

/** What a successful aggregator read yields — three raw hex payloads. */
export interface AggregatorRaw {
  latestRoundData: string
  decimals: string
  description: string
}

const sleep = (ms: number) => new Promise(resolve => setTimeout(resolve, ms))

/**
 * Aggregator constants, cached for the lifetime of the server process.
 *
 * Keyed by `chainId:address` rather than by network name, so two network entries that
 * happen to point at the same chain cannot cross-contaminate each other's cache.
 */
const constantsCache = new Map<string, { decimals: string; description: string }>()

export class RpcClient {
  readonly endpoint: Endpoint
  private readonly timeoutMs: number
  private readonly retries: number
  private requestId = 0

  constructor(endpoint: Endpoint, timeoutMs = DEFAULT_TIMEOUT_MS, retries = DEFAULT_RETRIES) {
    this.endpoint = endpoint
    this.timeoutMs = timeoutMs
    this.retries = Math.max(0, retries)
  }

  // ── transport ──────────────────────────────────────────────────────────────

  /** Exactly one HTTP round trip. Retry policy lives in `post`, not here. */
  private async attempt(body: string): Promise<unknown> {
    const controller = new AbortController()
    const timer = setTimeout(() => controller.abort(), this.timeoutMs)
    try {
      const response = await fetch(this.endpoint.url, {
        method: 'POST',
        headers: {
          'Content-Type': 'application/json',
          Accept: 'application/json',
          // Several public providers 403 an unrecognised agent outright, which reads
          // as "the chain is down" if you have not hit it before.
          'User-Agent': 'alchem-link-web',
        },
        body,
        signal: controller.signal,
        cache: 'no-store',
      })

      if (!response.ok) {
        const detail = (await response.text().catch(() => '')).slice(0, 200)
        // 4xx will not become 2xx on a retry; fail fast and say why. 429 is the
        // exception — backing off is exactly the right response there.
        const fatal = response.status >= 400 && response.status < 500 && response.status !== 429
        throw new RpcTransportError(
          `HTTP ${response.status} from ${redact(this.endpoint.url)}${detail ? `: ${detail}` : ''}`,
          !fatal,
        )
      }

      const text = await response.text()
      try {
        return JSON.parse(text)
      } catch {
        throw new RpcTransportError(`${redact(this.endpoint.url)} returned a non-JSON body`, false)
      }
    } catch (err) {
      if (err instanceof RpcTransportError) throw err
      const reason = err instanceof Error && err.name === 'AbortError'
        ? `timed out after ${this.timeoutMs}ms`
        : err instanceof Error ? err.message : String(err)
      throw new RpcTransportError(`${redact(this.endpoint.url)}: ${reason}`)
    } finally {
      clearTimeout(timer)
    }
  }

  private async post(body: string): Promise<unknown> {
    let lastError: unknown = null
    for (let attempt = 0; attempt <= this.retries; attempt++) {
      try {
        return await this.attempt(body)
      } catch (err) {
        if (err instanceof RpcTransportError && !err.retryable) throw err
        lastError = err
        if (attempt < this.retries) await sleep(250 * (attempt + 1))
      }
    }
    const detail = lastError instanceof Error ? lastError.message : String(lastError)
    throw new RpcTransportError(
      `could not reach ${redact(this.endpoint.url)} after ${this.retries + 1} attempt(s): ${detail}`,
    )
  }

  private static unwrap(envelope: RpcEnvelope, method: string): unknown {
    if (envelope?.error) {
      const { code, message } = envelope.error
      throw new RpcError(
        `${method} failed${code !== undefined ? ` [${code}]` : ''}: ${message ?? 'unknown error'}`,
      )
    }
    if (!envelope || !('result' in envelope)) {
      throw new RpcError(`${method} returned no result field`)
    }
    return envelope.result
  }

  async call(method: string, params: unknown[] = []): Promise<unknown> {
    const id = ++this.requestId
    const body = JSON.stringify({ jsonrpc: '2.0', id, method, params })
    return RpcClient.unwrap((await this.post(body)) as RpcEnvelope, method)
  }

  /**
   * Send several calls as one JSON-RPC batch, preserving request order in the result.
   *
   * Results are **settled, not thrown**. One aggregator reverting must not blank the
   * whole board — the Python reader skips a failing feed and returns the rest, and this
   * keeps that property. A transport failure still throws, because that one really did
   * take out every call in the batch.
   *
   * Responses may come back in any order, so they are re-keyed by id rather than zipped
   * positionally. A provider that reordered would otherwise hand you one feed's price
   * under another feed's label, which is the worst failure available here.
   */
  async callBatch(requests: RpcRequest[]): Promise<Settled[]> {
    if (requests.length === 0) return []

    const withIds = requests.map(req => ({
      jsonrpc: '2.0' as const,
      id: ++this.requestId,
      method: req.method,
      params: req.params ?? [],
    }))

    const body = await this.post(JSON.stringify(withIds))

    if (!Array.isArray(body)) {
      // Provider does not support batch. Fall back to individual calls rather than
      // failing the whole panel over a transport detail.
      return Promise.all(
        requests.map(async (req): Promise<Settled> => {
          try {
            return { ok: true, result: await this.call(req.method, req.params ?? []) }
          } catch (err) {
            if (err instanceof RpcTransportError) throw err
            return { ok: false, error: err instanceof Error ? err.message : String(err) }
          }
        }),
      )
    }

    const byId = new Map<number, RpcEnvelope>()
    for (const envelope of body as RpcEnvelope[]) byId.set(envelope.id, envelope)

    return withIds.map((sent): Settled => {
      const envelope = byId.get(sent.id)
      if (!envelope) return { ok: false, error: `${sent.method} missing from the batch response` }
      try {
        return { ok: true, result: RpcClient.unwrap(envelope, sent.method) }
      } catch (err) {
        return { ok: false, error: err instanceof Error ? err.message : String(err) }
      }
    })
  }

  // ── convenience wrappers ───────────────────────────────────────────────────

  async blockNumber(): Promise<number> {
    return parseInt((await this.call('eth_blockNumber')) as string, 16)
  }

  async chainId(): Promise<number> {
    return parseInt((await this.call('eth_chainId')) as string, 16)
  }

  async gasPriceWei(): Promise<bigint> {
    return BigInt((await this.call('eth_gasPrice')) as string)
  }

  ethCallRequest(to: string, data: string): RpcRequest {
    return { method: 'eth_call', params: [{ to, data }, 'latest'] }
  }

  /**
   * Fetch the three aggregator fields a price read needs, for many addresses at once.
   *
   * `decimals()`/`description()` are only requested for addresses not already cached, so
   * a steady-state poll costs one `latestRoundData()` per feed.
   *
   * Per-address outcome: either the three raw payloads, or the reason that one address
   * failed. The caller renders the failures instead of quietly dropping them.
   */
  async readAggregators(
    addresses: string[],
  ): Promise<Map<string, { ok: true; raw: AggregatorRaw } | { ok: false; error: string }>> {
    const chainId = this.endpoint.network.chainId
    const cacheKey = (address: string) => `${chainId}:${address.toLowerCase()}`

    const requests: RpcRequest[] = []
    const plan: { address: string; needsConstants: boolean }[] = []

    for (const address of addresses) {
      const cached = constantsCache.get(cacheKey(address))
      requests.push(this.ethCallRequest(address, SELECTOR_LATEST_ROUND_DATA))
      if (!cached) {
        requests.push(this.ethCallRequest(address, SELECTOR_DECIMALS))
        requests.push(this.ethCallRequest(address, SELECTOR_DESCRIPTION))
      }
      plan.push({ address, needsConstants: !cached })
    }

    const results = await this.callBatch(requests)

    const out = new Map<string, { ok: true; raw: AggregatorRaw } | { ok: false; error: string }>()
    let cursor = 0
    for (const { address, needsConstants } of plan) {
      const round = results[cursor++]
      const rawDecimals = needsConstants ? results[cursor++] : null
      const rawDescription = needsConstants ? results[cursor++] : null

      // Collect every failure for this address, so the panel can say which call broke.
      const failure = [round, rawDecimals, rawDescription].find(
        (r): r is { ok: false; error: string } => r !== null && !r.ok,
      )
      if (failure) {
        out.set(address, { ok: false, error: failure.error })
        continue
      }

      let constants = constantsCache.get(cacheKey(address))
      if (needsConstants) {
        constants = {
          decimals: (rawDecimals as { ok: true; result: unknown }).result as string,
          description: (rawDescription as { ok: true; result: unknown }).result as string,
        }
        // Only cached after a clean decode-able read, so a transient error never
        // becomes a permanently wrong `decimals` for the process lifetime.
        constantsCache.set(cacheKey(address), constants)
      }

      out.set(address, {
        ok: true,
        raw: {
          latestRoundData: (round as { ok: true; result: unknown }).result as string,
          ...constants!,
        },
      })
    }
    return out
  }
}

export function gwei(wei: bigint): number {
  return Number(wei) / 1e9
}
